### PR TITLE
feat(core): implement explain transaction method for Casper

### PR DESCRIPTION
**Pull request by fmeichtri-simtlix**
Implementing explainTransaction method for Casper, unitests were added for the implemented method.

Jira Ticket: [STLX-677](https://bitgoinc.atlassian.net/browse/STLX-677)

## Changes:
- explainTransaction method implementation
- unitests for explainTransaction

### CODE REVIEW

File: modules/core/src/v2/coins/cspr.ts [hunk 1]
```diff
@@ -31,10 +33,29 @@ export interface TransactionPrebuild extends BaseTransactionPrebuild {
   txJson: string;
 }
 
+export interface TransactionFee {
```
Comment (rliberal-simtlix) [inline modules/core/src/v2/coins/cspr.ts:36]: Add js doc

File: modules/core/src/v2/coins/cspr.ts [hunk 2]
```diff
@@ -31,10 +33,29 @@ export interface TransactionPrebuild extends BaseTransactionPrebuild {
...
+  gasLimit: string;
+  gasPrice: string;
+}
+
+export interface ExplainTransactionOptions {
+  txHex?: string;
```
Comment (rliberal-simtlix) [inline modules/core/src/v2/coins/cspr.ts:42]: add js doc

File: modules/core/src/v2/coins/cspr.ts [hunk 3]
```diff
@@ -31,10 +33,29 @@ export interface TransactionPrebuild extends BaseTransactionPrebuild {
...
+    txHex: string;
+  };
+  fee: TransactionFee;
+}
+
 interface SupplementGenerateWalletOptions {
   rootPrivateKey?: string;
 }
 
+interface TransactionOutput {
```
Comment (rliberal-simtlix) [inline modules/core/src/v2/coins/cspr.ts:53]: add js doc

File: modules/core/src/v2/coins/cspr.ts [hunk 4]
```diff
@@ -201,4 +222,63 @@ export class Cspr extends BaseCoin {
       .call(this)
       .asCallback(callback);
   }
+
+  /**
+   * Explain a Casper transaction from Raw Tx
+   *
+   * @param params
```
Comment (rliberal-simtlix) [inline modules/core/src/v2/coins/cspr.ts:229]: explain here list of params

File: modules/core/src/v2/coins/cspr.ts [hunk 5]
```diff
@@ -201,4 +222,63 @@ export class Cspr extends BaseCoin {
...
+   */
+  explainTransaction(
+    params: ExplainTransactionOptions,
+    callback?: NodeCallback<TransactionExplanation>
+  ): Bluebird<TransactionExplanation> {
+    const self = this;
+    return co<TransactionExplanation>(function*() {
+      const txHex = params.txHex || (params.halfSigned && params.halfSigned.txHex);
+      if (!txHex || !params.fee) {
+        throw new Error('missing explain tx parameters');
```
Comment (rliberal-simtlix) [inline modules/core/src/v2/coins/cspr.ts:240]: Nit: Missing

File: modules/core/src/v2/coins/cspr.ts [hunk 6]
```diff
@@ -201,4 +222,63 @@ export class Cspr extends BaseCoin {
...
+          address: toAddress,
+          amount,
+          coin: self.getChain(),
+        });
+      }
+      const outputAmount = outputs
+        .reduce((acumulator, output) => {
+          const currentValue = new BigNumber(output.amount);
+          return acumulator.plus(currentValue);
+        }, new BigNumber(0))
```
Comment (rliberal-simtlix) [inline modules/core/src/v2/coins/cspr.ts:265]: Explain those magic numbers

File: modules/core/src/v2/coins/cspr.ts (hunk 5 above)
Comment (rliberal-simtlix) [inline modules/core/src/v2/coins/cspr.ts:240]: There no an error for invalid params?

**Review by rliberal-simtlix — CHANGES_REQUESTED** [review]

File: modules/core/src/v2/coins/cspr.ts (hunk 5 above)
Comment (fmeichtri-simtlix) [reply modules/core/src/v2/coins/cspr.ts:240]: I don't think that there is a custom error for missing or invalid parameters. Taking as an example [XTZ](https://github.com/BitGo/BitGoJS/blob/5a30f69827aa636e767250c381983769d66273a6/modules/core/src/v2/coins/xtz.ts#L267) and [XRP](https://github.com/BitGo/BitGoJS/blob/5a30f69827aa636e767250c381983769d66273a6/modules/core/src/v2/coins/xrp.ts#L288).. as you can see the error messages are set starting with lower case too.

Comment (fmeichtri-simtlix) [reply modules/core/src/v2/coins/cspr.ts:240]: please see the comment https://github.com/BitGo/BitGoJS/pull/1066#discussion_r596169759. I think it's ok setting the message in lowercase

File: modules/core/src/v2/coins/cspr.ts (hunk 4 above)
Comment (fmeichtri-simtlix) [reply modules/core/src/v2/coins/cspr.ts:229]: You're right, I can improve this jsdoc. Thanks!

**Review by CapnMigraine — APPROVED** [review]
lgtm

File: modules/core/src/v2/coins/cspr.ts (hunk 6 above)
Comment (fmeichtri-simtlix) [reply modules/core/src/v2/coins/cspr.ts:269]: It's just the default accumulator value in the reduce. Here I'm defining the acumulator as a Big Number with value 0, e.g. if we don't have any outputs the returned value will be 0.

File: modules/core/src/v2/coins/cspr.ts (hunk 4 above)
Comment (fmeichtri-simtlix) [reply modules/core/src/v2/coins/cspr.ts]: Done!

**Review by rliberal-simtlix — APPROVED** [review]
LGTM

File: modules/core/src/v2/coins/cspr.ts [hunk 7]
```diff
@@ -31,10 +33,29 @@ export interface TransactionPrebuild extends BaseTransactionPrebuild {
...
+}
+
 interface SupplementGenerateWalletOptions {
   rootPrivateKey?: string;
 }
 
+interface TransactionOutput {
+  address: string;
+  amount: string;
+  coin: string;
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/cspr.ts:56]: How come each output has a coin? Is it possible to send a tx which spends more than one type of coin?

File: modules/core/src/v2/coins/cspr.ts [hunk 8]
```diff
@@ -201,4 +222,67 @@ export class Cspr extends BaseCoin {
...
+      const transferId = accountLib.Cspr.Utils.getTransferId(tx.casperTx.session);
+      const amount = accountLib.Cspr.Utils.getTransferAmount(tx.casperTx.session);
+      const toAddress = accountLib.Cspr.Utils.getTransferDestinationAddress(tx._deploy.session);
+      const outputs: TransactionOutput[] = [];
+      // TODO(https://bitgoinc.atlassian.net/browse/STLX-677): define outputs for different types of tx
+      if (tx.type === accountLib.BaseCoin.TransactionType.Send) {
+        outputs.push({
+          address: toAddress,
+          amount,
+          coin: self.getChain(),
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/cspr.ts:262]: Seems this is always set to the same value - is there some other reason this is needed?

File: modules/core/src/v2/coins/cspr.ts (hunk 7 above)
Comment (CapnMigraine) [reply modules/core/src/v2/coins/cspr.ts:56]: the coin will be used when we add support for token transactions

**Pull request merged**